File: SampleDirectXForms/D3DGame.cs
namespace SampleDirectXForms;

using System.Threading;

internal class D3DGame(nint instanceHandle)
    : D3DWindow(instanceHandle)
{
    private readonly D3DRenderer renderer = new();
    private readonly GameTimer timer = new();
    private float frameCount;
    private bool isPaused;
    private float timeElapsed;

    public override void Init()
    {
        base.Init();
        this.renderer.Init();
        this.timer.Reset();
    }

    public void Tick()
    {
        this.timer.Tick();

        if (this.isPaused)
        {
            Thread.Sleep(100);
            return;
        }

        this.CalculateFrameStats();
        this.UpdateScene(this.timer.DeltaTime);
        this.DrawScene();

        Thread.Sleep(0);
    }

    protected virtual void DrawScene()
    {
    }

    protected override void OnActivated()
    {
        this.isPaused = false;
        this.timer.Start();
    }

    protected override void OnDeactivated()
    {
        this.isPaused = true;
        this.timer.Stop();
    }

    protected override void OnResize()
    {
        this.isPaused = false;
        this.timer.Start();
    }

    protected override void OnResizing()
    {
        this.isPaused = true;
        this.timer.Stop();
    }

    protected virtual void UpdateScene(float dt)
    {
    }

    private void CalculateFrameStats()
    {
        this.frameCount++;

        var gameTime = this.timer.GameTime;

        if (gameTime - this.timeElapsed < 1.0f)
        {
            return;
        }

        var fps = (float)this.frameCount;
        var frameTime = 1000.0f / fps;

        var text = $"{this.Caption} FPS: {fps:0} Frame Time: {frameTime:0.000000} Game Time: {gameTime:0}";
        NativeMethods.SetWindowText(this.Handle, text);

        this.frameCount = 0;
        this.timeElapsed += 1.0f;
    }
}

File: SampleDirectXNative/D3DGame.cpp
#include "D3DGame.h"

D3DGame::D3DGame(HINSTANCE hInstance) :
    D3DWindow(hInstance),
    mAppPaused(false)
{
}

D3DGame::~D3DGame()
{
}

bool D3DGame::Init()
{
    D3DWindow::Init();
    mRenderer.Init();
    mTimer.Reset();

    return true;
}

void D3DGame::Tick()
{
    mTimer.Tick();

    if (mAppPaused)
    {
        Sleep(100);
        return;
    }
    
    CalculateFrameStats();
    UpdateScene(mTimer.DeltaTime());
    DrawScene();

    Sleep(0);
}

void D3DGame::OnActivated()
{
    mAppPaused = false;
    mTimer.Start();
}

void D3DGame::OnDeactivated()
{
    mAppPaused = true;
    mTimer.Stop();
}

void D3DGame::OnResize()
{
    mAppPaused = false;
    mTimer.Start();
}

void D3DGame::OnResizing()
{
    mAppPaused = true;
    mTimer.Stop();
}

void D3DGame::CalculateFrameStats() const
{
    static int frameCnt = 0;
    static float timeElapsed = 0.0f;
    frameCnt++;

    const float gameTime = mTimer.GameTime();

    if (gameTime - timeElapsed < 1.0f)
    {
        return;
    }

    const float fps = static_cast<float>(frameCnt);
    const float mspf = 1000.0f / fps;

    std::wostringstream outs;
    outs.precision(6);
    outs << mMainWndCaption << L" "
        << L"FPS: " << fps << L" "
        << L"Frame Time: " << mspf << L" (ms) "
        << L"Game Time: " << gameTime << L" (s)";

    SetWindowText(mhMainWnd, outs.str().c_str());

    frameCnt = 0;
    timeElapsed += 1.0f;
}
